crates/iota-data-ingestion-core/src/tests.rs

```diff
@@ -253,6 +277,45 @@ async fn worker_pool_with_reducer() {
     assert_eq!(result.unwrap().get("test"), Some(&20));
 }
 
+// Tests the graceful shutdown behavior when reducer encounter persistent
+// failures.
+//
+// This test verifies that:
+// 1. When Reducer::commit implementation continuously fails.
+// 2. The exponential backoff retry mechanism would normally create a loop until
+//    the successful value is returned.
+// 3. The graceful shutdown logic successfully breaks these retry loops upon
+//    cancellation.
+// 4. The Reducer exit cleanly without committing any batch.
+//
+// The test uses `FaultyReducer` which always fails, simulating a worst-case
+// scenario where all WorkerPools are unable to send progress data to
+// IndexerExecutor.
+#[tokio::test]
+async fn graceful_shutdown_faulty_reducer() {
```

Comment:
Would it make sense to also have a test to graceful shutdown non-faulty reducer?

Reply:
The `worker_pool_with_reducer` test case does test the graceful shutdown with a non faulty reducer